connector/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryDataWriterHelper.java

```diff
@@ -0,0 +1,364 @@
+      throw new RuntimeException("Could not retrieve AppendRowsResponse.", e);
+    }
+    if (appendRowsResponse.hasError()) {
+      throw new IOException(
+          "Append request failed with error: " + appendRowsResponse.getError().getMessage());
+    }
+    long responseOffset = appendRowsResponse.getOffset();
+    if (expectedOffset != responseOffset) {
+      throw new IOException(
+          String.format(
+              "On stream %s append-rows response offset %d did not match expected offset %d.",
+              writeStreamName, responseOffset, expectedOffset));
+    }
+  }
+
+  /**
+   * Appends any data that remains in the protoRows, waits for 500 milliseconds, and finalizes the
+   * write-stream.
+   *
+   * @return The finalized row-count of the write-stream.
+   * @throws IOException If the row-count returned by the FinalizeWriteStreamResponse does not match
+   *     the expected offset (which is equal to the number of rows appended thus far).
+   * @see this#writeStreamRowCount
+   */
+  @Override
+  public long commit() throws IOException {
+    if (this.protoRows.getSerializedRowsCount() != 0) {
+      sendAppendRowsRequest();
+    }
+
+    waitBeforeFinalization();
+
+    Storage.FinalizeWriteStreamRequest finalizeWriteStreamRequest =
+        Storage.FinalizeWriteStreamRequest.newBuilder().setName(writeStreamName).build();
+    Storage.FinalizeWriteStreamResponse finalizeResponse =
+        retryFinalizeWriteStream(finalizeWriteStreamRequest);
+
+    long expectedFinalizedRowCount = writeStreamRowCount;
+    long responseFinalizedRowCount = finalizeResponse.getRowCount();
+    if (responseFinalizedRowCount != expectedFinalizedRowCount) {
+      throw new IOException(
+          String.format(
+              "On stream %s finalization, expected finalized row count %d but received %d",
+              writeStreamName, expectedFinalizedRowCount, responseFinalizedRowCount));
+    }
+
+    writeClient.shutdown();
+
+    logger.debug(
+        "Write-stream {} finalized with row-count {}", writeStreamName, responseFinalizedRowCount);
+
+    return responseFinalizedRowCount;
+  }
+
+  /**
+   * A helper method in order to submit for retry (using method retryCallable) a callable that
+   * finalizes the write-stream; useful if finalization encountered a retriable error.
+   *
+   * @param finalizeWriteStreamRequest The request to send to the writeClient in order to finalize
+   *     the write-stream.
+   * @return The FinalizeWriteStreamResponse
+   * @see com.google.cloud.bigquery.storage.v1alpha2.Storage.FinalizeWriteStreamResponse
+   */
+  private Storage.FinalizeWriteStreamResponse retryFinalizeWriteStream(
+      Storage.FinalizeWriteStreamRequest finalizeWriteStreamRequest) {
+    try {
+      return retryCallable(() -> writeClient.finalizeWriteStream(finalizeWriteStreamRequest));
+    } catch (ExecutionException | InterruptedException e) {
+      throw new RuntimeException(
```

Comment:
Please change RuntimeException to a subclass with better description
nit: remove dot in the end of the message